rfcs/syntax-safe-navigation-operator.md

```diff
@@ -0,0 +1,137 @@
+# Safe navigation postfix operator (?)

```

Comment:
Since the operator cannot be used without `.` and `:` is it even a standalone operator `?` or are these really operators `?.` and `?:`?

For example, C# calls them `Null-conditional operators ?. and ?[]` and that seems to align better with the behavior that is described here (but ? and . are still separate tokens)

Reply:
Wouldn't `?` just be sugar for `if x then x... else nil`? It feels like we could support them and have this be its own operator.

Reply:
@vegorov-rbx It's probably just simpler documentation-wise to say that `?.` and `?[]` is a whole new operator than to explain the internal details of how `?` actually works from the parser/codegen standpoint. I imagine that we'd do the same in our documentation.

@BradSharp I don't think that would be a 1:1 sugar. You don't want redundant calls to `__index` for every `?.`/?:`/`?[` in the expression path.

```lua
local print_index_mt = {
  __index = function(self, key)
    print(key)
    return rawget(self, "_" .. key)
  end
}

local t = setmetatable({
  _p1 = setmetatable({ _p2 = 5 }, print_index_mt)
}, print_index_mt)

print(t?.p1?.p2)
-- Should print the following:
--> p1
--> p2
--> 5
--
-- And not:
--> p1
--> p1
--> p2
--> 5
```

Reply:
@alexmccord ah yes I didn't consider that. A more accurate version would be:
```lua
local _temp = <lhs>
<result> = if _temp then _temp<rhs> else nil
```